core/server/common/src/main/java/alluxio/underfs/UfsManager.java

```diff
@@ -0,0 +1,167 @@
+      if (object == this) {
+        return true;
+      }
+
+      if (!(object instanceof Key)) {
+        return false;
+      }
+
+      Key that = (Key) object;
+      return Objects.equal(mAuthority, that.mAuthority) && Objects
+          .equal(mProperties, that.mProperties) && Objects.equal(mScheme, that.mScheme);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+          .add("authority", mAuthority)
+          .add("scheme", mScheme)
+          .add("properties", mProperties).toString();
+    }
+  }
+
+  /** Maps from key to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Key, UnderFileSystem> mUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+  /** Maps from mount id to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Long, UnderFileSystem> mMountIdToUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+
+  private final UnderFileSystem mRootUfs;
+  protected final Closer mCloser;
+
+  /**
+   * Constructs the instance.
+   */
+  public UfsManager() {
+    mCloser = Closer.create();
+    String rootUri = Configuration.get(PropertyKey.MASTER_MOUNT_TABLE_ROOT_UFS);
+    Map<String, String> rootConf =
+        Configuration.getNestedProperties(PropertyKey.MASTER_MOUNT_TABLE_ROOT_OPTION);
+    mRootUfs = UnderFileSystemRegistry.create(rootUri, rootConf);
+    mUnderFileSystemMap.put(new Key(new AlluxioURI(rootUri), rootConf), mRootUfs);
+    mMountIdToUnderFileSystemMap.put(IdUtils.ROOT_MOUNT_ID, mRootUfs);
+    mCloser.register(mRootUfs);
+  }
+
+  /**
+   * Gets a UFS instance from the cache if exists. Otherwise, creates a new instance and adds
+   * that to the cache. Use this method only when you create new UFS instances.
+   *
+   * @param path the UFS path
+   * @param ufsConf the UFS configuration
+   * @return the UFS instance
+   */
+  public UnderFileSystem getOrCreate(String path, Map<String, String> ufsConf) {
+    Key key = new Key(new AlluxioURI(path), ufsConf);
+    UnderFileSystem cachedFs = mUnderFileSystemMap.get(key);
+    if (cachedFs != null) {
+      return cachedFs;
+    }
+    UnderFileSystem fs = UnderFileSystemRegistry.create(path, ufsConf);
+    UnderFileSystem racingFs = mUnderFileSystemMap.putIfAbsent(key, fs);
+    if (racingFs == null) {
+      mCloser.register(fs);
+      return fs;
+    }
+    try {
+      fs.close();
+    } catch (IOException e) {
+      // Cannot close the created ufs which fails the race.
+      LOG.error("Failed to close ufs {}", fs, e);
+    }
+    return racingFs;
+  }
+
+  /**
+   * Marks a UFS instance as a mount entry and associates it with a mount id.
+   *
+   * @param ufs the UFS
+   * @param mountId the mount id
+   */
+  public void addMount(UnderFileSystem ufs, long mountId) {
+    Preconditions.checkArgument(ufs != null, "ufs");
+    Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
+    mMountIdToUnderFileSystemMap.put(mountId, ufs);
```

Comment:
- I combined `addMount` and `getOrCreate`, meanwhile added a `removeMount`. So whenever we add a new instance to `mUnderFileSystemMap`, it will be indexed by `mountId`. In other words, the two maps pointing to the same set of UFS instances.
- Doc added